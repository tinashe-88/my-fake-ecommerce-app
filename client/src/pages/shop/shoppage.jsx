import React, { useEffect, lazy, Suspense } from 'react'
import { Route } from 'react-router-dom'
import { connect } from 'react-redux'

import { fetchCollectionsStart } from '../../redux/shop/shop.actions'
import LoadingSpinner from '../../components/loading-spinner/loading-spinner'

import { ShopPageContainer } from './shoppage.styles'

const CollectionPageContainer = lazy(() => 
  import('../../pages/collection/collection.container')
)
const CollectionsOverviewContainer = lazy(() => 
  import('../../components/collections-overview/collections-overview')
)

export const ShopPage = ({ fetchCollectionsStart, match }) => {
  useEffect(() => { 
    fetchCollectionsStart()
  }, [fetchCollectionsStart])
   
  return (
    <ShopPageContainer>
      <Suspense fallback={<LoadingSpinner />}>
        <Route 
          exact 
          path={`${match.path}`} 
          component={CollectionsOverviewContainer}
        />
        <Route 
          path={`${match.path}/:collectionId`}
          component={CollectionPageContainer}
        />
      </Suspense>
    </ShopPageContainer>
  )
}

const mapDispatchToProps = dispatch => ({
  fetchCollectionsStart: () => dispatch(fetchCollectionsStart())
})

export default connect(null, mapDispatchToProps)(ShopPage)