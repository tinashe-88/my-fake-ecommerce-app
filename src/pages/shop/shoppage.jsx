import React from 'react'
import { Route } from 'react-router-dom'
import { connect } from 'react-redux'

import { fetchCollectionsStartAsync } from '../../redux/shop/shop.actions'

import CollectionPageContainer from '../../pages/collection/collection.container'
import CollectionsOverviewContainer from '../../components/collections-overview/collections-overview'

class ShopPage extends React.Component {
  componentDidMount(){
    const { fetchCollectionsStartAsync } = this.props

    fetchCollectionsStartAsync()
  }

  render(){
    const { match } = this.props

    return (
      <>
        <Route 
          exact 
          path={`${match.path}`} 
          component={CollectionsOverviewContainer}
        />
        <Route 
          path={`${match.path}/:collectionId`}
          component={CollectionPageContainer}
        />
      </>
    )
  }
}

const mapDispatchToProps = dispatch => ({
  fetchCollectionsStartAsync: () => dispatch(fetchCollectionsStartAsync())
})

export default connect(null, mapDispatchToProps)(ShopPage)