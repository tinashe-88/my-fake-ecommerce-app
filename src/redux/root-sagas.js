import { all, call } from 'redux-saga/effects'

import { shopSagas } from './shop/shop.sagas'

function* rootSaga(){
    yield all([
        call(shopSagas)
    ])
}

export default rootSaga