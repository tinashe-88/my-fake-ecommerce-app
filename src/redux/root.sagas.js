import { all, call } from 'redux-saga/effects'

import { userSagas } from './user/user.sagas'
import { fetchCollectionsStart } from './shop/shop.sagas'

function* rootSaga(){
    yield all([
        call(userSagas),
        call(fetchCollectionsStart)
    ])
}

export default rootSaga