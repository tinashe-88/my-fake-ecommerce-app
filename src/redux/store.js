import {
  createStore,
  applyMiddleware
} from 'redux'
import { persistStore } from 'redux-persist'
import logger from 'redux-logger'
import createSagaMiddleware from 'redux-saga'

import rootReducer from './root.reducer'

import { fetchCollectionsStart } from './shop/shop.sagas'

const sagaMiddleware = createSagaMiddleware()

const middleware = [sagaMiddleware]

if (process.env.NODE_ENV === 'development'){
  middleware.push(logger)
}

export const store = createStore(rootReducer, applyMiddleware(...middleware))

sagaMiddleware.run(fetchCollectionsStart)

export const persistor = persistStore(store)