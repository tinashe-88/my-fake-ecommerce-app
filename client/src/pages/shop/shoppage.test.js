import React from 'react'
import { mount } from 'enzyme'
import { combineReducers, createStore } from 'redux'
import { BrowserRouter } from 'react-router-dom'
import { Provider } from 'react-redux'

import { ShopPage } from './shoppage'

export const createMockStore = ({ state, reducers}) => {
  const store = createStore(combineReducers(reducers), state)

  return {
    ...store,
    persistor: {
      persist: () => null
    }
  }
}

describe('ShopPage component', () => {
  let wrapper
  let mockFetchCollectionsStart
  let store

  beforeEach(() => {
    const mockReducer = (
      state = {
        isFetching: true
      },
      action
    ) => state
    

    const mockState = {
      shop: { 
        isFetching: true
      }
    }

    mockFetchCollectionsStart = jest.fn()

    store = createMockStore({
      state: mockState,
      reducers: {
        shop: mockReducer
      }
    })

    const mockMatch = {
      path: ''
    }

    const mockProps = {
      match: mockMatch,
      fetchCollectionsStart: mockFetchCollectionsStart
    }

    wrapper = mount(
      <BrowserRouter>
        <Provider store={store}>
          <ShopPage {...mockProps}/>
        </Provider>
      </BrowserRouter>
    )
  })

  it('Should render ShopPage component', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Should render ShopPage component', () => {
    expect(mockFetchCollectionsStart).toHaveBeenCalled()
  })
})