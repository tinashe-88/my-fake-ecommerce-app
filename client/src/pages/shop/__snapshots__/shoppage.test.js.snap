// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopPage component Should render ShopPage component 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "dispatch": [Function],
          "getState": [Function],
          "persistor": Object {
            "persist": [Function],
          },
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <ShopPage
        fetchCollectionsStart={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        match={
          Object {
            "path": "",
          }
        }
      >
        <styled.div>
          <div
            className="sc-AxiKw hSLCUF"
          >
            <Suspense
              fallback={<LoadingSpinner />}
            >
              <Route
                component={
                  Object {
                    "$$typeof": Symbol(react.lazy),
                    "_ctor": [Function],
                    "_result": Object {
                      "$$typeof": Symbol(react.memo),
                      "WrappedComponent": [Function],
                      "compare": null,
                      "displayName": "Connect(CollectionsOverview)",
                      "type": [Function],
                    },
                    "_status": 1,
                  }
                }
                exact={true}
                path=""
              />
              <Route
                component={
                  Object {
                    "$$typeof": Symbol(react.lazy),
                    "_ctor": [Function],
                    "_result": null,
                    "_status": -1,
                  }
                }
                path="/:collectionId"
              />
              <LoadingSpinner>
                <styled.div>
                  <div
                    className="sc-AxjAm HvuBM"
                  >
                    <styled.div>
                      <div
                        className="sc-AxirZ eovtYN"
                      />
                    </styled.div>
                  </div>
                </styled.div>
              </LoadingSpinner>
            </Suspense>
          </div>
        </styled.div>
      </ShopPage>
    </Provider>
  </Router>
</BrowserRouter>
`;
